iroh-blobs/src/util/local_pool.rs

```diff
@@ -0,0 +1,698 @@
+    }
+}
+
+impl From<SpawnError> for std::io::Error {
+    fn from(e: SpawnError) -> Self {
+        std::io::Error::new(std::io::ErrorKind::Other, e)
+    }
+}
+
+impl LocalPoolHandle {
+    /// Get the number of tasks in the queue
+    ///
+    /// This is *not* the number of tasks being executed, but the number of
+    /// tasks waiting to be scheduled for execution. If this number is high,
+    /// it indicates that the pool is very busy.
+    ///
+    /// You might want to use this to throttle or reject requests.
+    pub fn waiting_tasks(&self) -> usize {
+        self.send.len()
+    }
+
+    /// Spawn a new task and return a tokio join handle.
+    ///
+    /// This is like [`LocalPoolHandle::spawn_pinned`], but does not panic if
+    /// the pool is shut down.
+    pub fn try_spawn_pinned<T, F, Fut>(&self, gen: F) -> SpawnResult<tokio::task::JoinHandle<T>>
+    where
+        F: FnOnce() -> Fut + Send + 'static,
+        Fut: Future<Output = T> + 'static,
+        T: Send + 'static,
+    {
+        let inner = self.try_run(gen)?;
+        let abort: Arc<OnceLock<AbortHandle>> = Arc::new(OnceLock::new());
+        let abort2 = abort.clone();
+        let res = tokio::spawn(async move {
+            match inner.await {
+                Ok(res) => res,
+                Err(_) => {
+                    // abort the outer task and wait forever (basically return pending)
+                    if let Some(abort) = abort.get() {
+                        abort.abort();
+                    }
+                    futures_lite::future::pending().await
+                }
+            }
+        });
+        let _ = abort2.set(res.abort_handle());
+        Ok(res)
+    }
+
+    /// Run a task in the pool and await the result.
+    ///
+    /// When the returned future is dropped, the task will be immediately
+    /// cancelled. Any drop implementation is guaranteed to run to completion in
+    /// any case.
+    pub fn try_run<T, F, Fut>(&self, gen: F) -> SpawnResult<Run<T>>
+    where
+        F: FnOnce() -> Fut + Send + 'static,
+        Fut: Future<Output = T> + 'static,
+        T: Send + 'static,
+    {
+        let (mut send_res, recv_res) = tokio::sync::oneshot::channel();
+        let item = move || async move {
+            let fut = (gen)();
+            tokio::select! {
+                // send the result to the receiver
+                res = fut => { send_res.send(res).ok(); }
+                // immediately stop the task if the receiver is dropped
+                _ = send_res.closed() => {}
+            }
+        };
+        self.try_spawn(item)?;
+        Ok(Run(recv_res))
+    }
+
+    /// Run a task in the pool.
+    ///
+    /// The task will be run detached. This can be useful if
+    /// you are not interested in the result or in in cancellation or
+    /// you provide your own result handling and cancellation mechanism.
+    pub fn try_spawn<F, Fut>(&self, gen: F) -> SpawnResult<()>
```

Comment:
maybe `try_spawn_detached`?

Reply:
doesn't spawn imply detached? At least that is how it is usually used in tokio. I was going for "run": you get back a handle, and dropping it will cancel the task, like our AbortableJoinHandle, and "spawn" for: you don't care about the result at this location.

Reply:
Ok, renamed it

(try_)run -> (try_)spawn
(try_)spawn -> (try_)spawn_detached